lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -1472,81 +1472,97 @@ bool MergeMatMul::run(Function *F, const CompilationContext &cctx) {
   return changed;
 }
 
-bool MergePadIntoConvolution::run(Function *F, const CompilationContext &cctx) {
-  LOG_SCOPE(F->getLogContext(), getName());
-  bool changed = false;
-  for (auto &node : F->getNodes()) {
-    auto *CN = dyn_cast<ConvolutionNode>(&node);
-    if (!CN) {
-      continue;
-    }
+/// Local utility to check whether the given convolution or pooling \p node has
+/// a Pad node for its input and tries to merge the Pad into the node.
+/// \returns true or false whether the pad was merged or not.
+template <class NodeTy> static bool mergePadIntoConvPool(Node *node) {
 
-    auto *PN = dyn_cast<PadNode>(CN->getInput());
-    if (!PN) {
-      continue;
-    }
+  auto *CN = dyn_cast<NodeTy>(node);
+  if (!CN) {
+    return false;
+  }
 
-    // Convolution only supports padding with 0 constant
-    if ((PN->getMode() != PaddingMode::CONSTANT) || (PN->getValue() != 0.f)) {
-      continue;
-    }
+  auto *PN = dyn_cast<PadNode>(CN->getInput());
+  if (!PN) {
+    return false;
+  }
 
-    // The convolution needs to be the unique user
-    if (!PN->hasOneUse()) {
-      continue;
-    }
+  // Convolution only supports padding with 0 constant
+  if ((PN->getMode() != PaddingMode::CONSTANT) || (PN->getValue() != 0.f)) {
+    return false;
+  }
 
-    // Compute the new padding.
-    // Note: - convolution only supports positive padding
-    //       - the convolution takes NHWC input tensors.
-    bool canMerge = true;
-    auto padPads = PN->getPads();
-    auto convPads = CN->getPads();
+  // The convolution needs to be the unique user
+  if (!PN->hasOneUse()) {
```

Comment:
Generally speaking, this is a redundant check. Pad can have multiple users, including the case where all of them are either Conv or Pool and we would be able to optimize Pad out by merging it with all of them.